app/serializers/osf-serializer.ts

```diff
@@ -1,96 +1,110 @@
-        },
     },
 
-    // Map from relationship field name to type. Override to serialize relationships.
-    relationshipTypes: {},
-
     /**
-     * Extract information about records embedded inside this request
+     * Get embedded objects from the response and push them into the store.
+     * Return a new resource hash that only contains relationships in JSON API format,
+     * or the original resource hash, unchanged.
+     *
      * @method _extractEmbeds
-     * @param {Object} resourceHash
-     * @return {Array}
+     * @param {ResourceHash} resourceHash
+     * @return {ResourceHash}
      * @private
      */
-
-    // TODO: refactor using resource hash like how we are using it here:
-    // (stuff like _mergeFields should be as easy as just assigning a copy object)
-
-    /* eslint-disable no-param-reassign */
-    _extractEmbeds(resourceHash: ResourceHash): any {
+    _extractEmbeds(this: OsfSerializer, resourceHash: ResourceHash): ResourceHash {
         if (!resourceHash.embeds) {
-            return []; // Nothing to do
+            return resourceHash;
         }
-        let included = [];
-        resourceHash.relationships = resourceHash.relationships || {};
-        for (const embedded of Object.keys(resourceHash.embeds)) {
-            if (!(embedded || resourceHash.embeds[embedded])) {
+        const relationships = {};
+        for (const relName of Object.keys(resourceHash.relationships)) {
+            const embeddedObj = resourceHash.embeds[relName];
+
+            if (!embeddedObj) {
+                // Non-embedded relationship; let it pass through unchanged
+                relationships[relName] = resourceHash.relationships[relName];
                 continue;
             }
-            // TODO Pagination probably breaks here
-            const data = resourceHash.embeds[embedded].data || resourceHash.embeds[embedded];
-            if (!('errors' in data)) {
-                this.store.pushPayload({ data });
+            if ('errors' in embeddedObj) {
+                // Should already be logged elsewhere
+                continue;
             }
-            if (Array.isArray(data)) {
-                included = included.concat(data);
+
+            // Push the embedded data into the store.  Since it will use this serializer again,
+            // this is indirectly recursive and will extract nested embeds.
+            this.get('store').pushPayload(embeddedObj);
+
+            // Merge links on the embedded object with links on the relationship, so all returned links are available
+            const embeddedLinks = Object.assign(
+                {},
+                embeddedObj.links,
+                resourceHash.relationships[relName].links,
+            );
+
+            // Construct a new relationship in JSON API format
+            if (Array.isArray(embeddedObj.data)) {
+                relationships[relName] = {
+                    data: embeddedObj.data.map(o => ({ id: o.id, type: o.type })),
```

Comment:
Can use destructuring and then take advantage of shorthand property names

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#Description